getSplitPointSlow unexpected split_handle

https://internal.pingcap.net/jira/browse/ONCALL-2870

```
[2021/04/22 17:51:59.154 +08:00] [ERROR] [<unknown>] ["void DB::ApplyPreHandledSnapshot(DB::TiFlashServer*, DB::PreHandledSnapshot*): Code: 0, e.displayText() = DB::Exception: getSplitPointSlow unexpected split_handle: 0, should be in range [549176901,552765639), e.what() = DB::Exception, Stack trace:
\n
\n0. bin/tiflash/tiflash(StackTrace::StackTrace()+0x15) [0x35a7aa5]
\n1. bin/tiflash/tiflash(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x25) [0x359e635]
\n2. bin/tiflash/tiflash(DB::DM::Segment::getSplitPointSlow(DB::DM::DMContext&, DB::DM::Segment::ReadInfo const&, std::shared_ptr<DB::DM::SegmentSnapshot> const&) const+0x747) [0x706a4e7]
\n3. bin/tiflash/tiflash(DB::DM::Segment::prepareSplitPhysical(DB::DM::DMContext&, std::shared_ptr<std::vector<DB::DM::ColumnDefine, std::allocator<DB::DM::ColumnDefine> > > const&, std::shared_ptr<DB::DM::SegmentSnapshot> const&, DB::DM::WriteBatches&) const+0x11e) [0x706e63e]
\n4. bin/tiflash/tiflash(DB::DM::Segment::prepareSplit(DB::DM::DMContext&, std::shared_ptr<std::vector<DB::DM::ColumnDefine, std::allocator<DB::DM::ColumnDefine> > > const&, std::shared_ptr<DB::DM::SegmentSnapshot> const&, DB::DM::WriteBatches&) const+0x98) [0x706f528]
\n5. bin/tiflash/tiflash(DB::DM::DeltaMergeStore::segmentSplit(DB::DM::DMContext&, std::shared_ptr<DB::DM::Segment> const&)+0x4ed) [0x704117d]
\n6. bin/tiflash/tiflash(DB::DM::DeltaMergeStore::checkSegmentUpdate(std::shared_ptr<DB::DM::DMContext> const&, std::shared_ptr<DB::DM::Segment> const&, DB::DM::DeltaMergeStore::ThreadType)+0x884) [0x7044a34]
\n7. bin/tiflash/tiflash(DB::DM::DeltaMergeStore::deleteRange(DB::Context const&, DB::Settings const&, DB::DM::Range<long> const&)+0x6f7) [0x70468c7]
\n8. bin/tiflash/tiflash(DB::StorageDeltaMerge::deleteRange(DB::DM::Range<long> const&, DB::Settings const&)+0x51) [0x6fc05b1]
\n9. bin/tiflash/tiflash(DB::KVStore::tryApplySnapshot(DB::RegionPtrWrap const&, DB::Context&)+0x581) [0x759b921]
\n10. bin/tiflash/tiflash(DB::KVStore::handleApplySnapshot(DB::RegionPtrWrap const&, DB::TMTContext&)+0x138) [0x759c888]
\n11. bin/tiflash/tiflash(DB::ApplyPreHandledSnapshot(DB::TiFlashServer*, DB::PreHandledSnapshot*)+0x3f) [0x721c38f]
\n12. bin/tiflash/tiflash(DB::ApplyPreHandledSnapshot(DB::TiFlashServer*, void*, DB::RawCppPtrType)+0x1d) [0x721d09d]
\n13. /work/tidb23400/bin/tiflash/libtiflash_proxy.so(+0x9e5849) [0x7fae2042c849]
\n14. /work/tidb23400/bin/tiflash/libtiflash_proxy.so(+0xa03e90) [0x7fae2044ae90]
\n15. /work/tidb23400/bin/tiflash/libtiflash_proxy.so(+0x496474) [0x7fae1fedd474]
\n16. /work/tidb23400/bin/tiflash/libtiflash_proxy.so(+0x4e8b4a) [0x7fae1ff2fb4a]
\n17. /work/tidb23400/bin/tiflash/libtiflash_proxy.so(+0x13663ce) [0x7fae20dad3ce]
\n18. /work/tidb23400/bin/tiflash/libtiflash_proxy.so(+0x137ec9b) [0x7fae20dc5c9b]
\n19. /lib64/libpthread.so.0(+0x7dc4) [0x7fae1f20ddc4]
\n20. /lib64/libc.so.6(clone+0x6c) [0x7fae1ec3a73c]
\n"] [thread_id=9]
```

--- a/dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp
+++ b/dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp
@@ -1162,13 +1162,17 @@ SegmentPair DeltaMergeStore::segmentSplit(DMContext & dm_context, const SegmentP
     });
 
     WriteBatches wbs(storage_pool);
-    auto         range      = segment->getRange();
-    auto         split_info = segment->prepareSplit(dm_context, schema_snap, segment_snap, wbs);
-    if (!split_info.ok)
+
+    auto range          = segment->getRange();
+    auto split_info_opt = segment->prepareSplit(dm_context, schema_snap, segment_snap, wbs);
+
+    if (!split_info_opt.has_value())
     {
         LOG_WARNING(log, "Give up segment [" << segment->segmentId() << "] split because of prepare split failed");
         return {};
     }
+
+    auto & split_info = split_info_opt.value();
 
     wbs.writeLogAndData();
     split_info.my_stable->enableDMFilesGC();

--- a/dbms/src/Storages/DeltaMerge/Segment.cpp
+++ b/dbms/src/Storages/DeltaMerge/Segment.cpp
@@ -541,9 +541,11 @@ SegmentPair Segment::split(DMContext & dm_context, const ColumnDefinesPtr & sche
     if (!segment_snap)
         return {};
 
-    auto split_info = prepareSplit(dm_context, schema_snap, segment_snap, wbs);
-    if (!split_info.ok)
+    auto split_info_opt = prepareSplit(dm_context, schema_snap, segment_snap, wbs);
+    if (!split_info_opt.has_value())
         return {};
+
+    auto & split_info = split_info_opt.value();
 
     wbs.writeLogAndData();
     split_info.my_stable->enableDMFilesGC();
@@ -699,10 +701,10 @@ Segment::getSplitPointSlow(DMContext & dm_context, const ReadInfo & read_info, c
     return {split_handle};
 }
 
-Segment::SplitInfo Segment::prepareSplit(DMContext &                dm_context,
-                                         const ColumnDefinesPtr &   schema_snap,
-                                         const SegmentSnapshotPtr & segment_snap,
-                                         WriteBatches &             wbs) const
+std::optional<Segment::SplitInfo> Segment::prepareSplit(DMContext &                dm_context,
+                                                        const ColumnDefinesPtr &   schema_snap,
+                                                        const SegmentSnapshotPtr & segment_snap,
+                                                        WriteBatches &             wbs) const
 {
     if (!dm_context.enable_logical_split         //
         || segment_snap->stable->getPacks() <= 3 //
@@ -712,25 +714,27 @@ Segment::SplitInfo Segment::prepareSplit(DMContext &                dm_context,
     }
     else
     {
-        auto split_point     = getSplitPointFast(dm_context, segment_snap->stable);
-        bool bad_split_point = !split_point.has_value() || !range.check(split_point.value()) || split_point == range.start;
+        auto split_point_opt = getSplitPointFast(dm_context, segment_snap->stable);
+        bool bad_split_point
+            = !split_point_opt.has_value() || !range.check(split_point_opt.value()) || split_point_opt.value() == range.start;
         if (bad_split_point)
         {
-            LOG_INFO(log,
-                     "Got bad split point [" << (split_point.has_value() ? Redact::handleToDebugString(split_point.value()) : "no value")
-                                             << "] for segment " << info() << ", fall back to split physical.");
+            LOG_INFO(
+                log,
+                "Got bad split point [" << (split_point_opt.has_value() ? Redact::handleToDebugString(split_point_opt.value()) : "no value")
+                                        << "] for segment " << info() << ", fall back to split physical.");
             return prepareSplitPhysical(dm_context, schema_snap, segment_snap, wbs);
         }
         else
-            return prepareSplitLogical(dm_context, schema_snap, segment_snap, split_point.value(), wbs);
+            return prepareSplitLogical(dm_context, schema_snap, segment_snap, split_point_opt.value(), wbs);
     }
 }
 
-Segment::SplitInfo Segment::prepareSplitLogical(DMContext & dm_context,
-                                                const ColumnDefinesPtr & /*schema_snap*/,
-                                                const SegmentSnapshotPtr & segment_snap,
-                                                Handle                     split_point,
-                                                WriteBatches &             wbs) const
+std::optional<Segment::SplitInfo> Segment::prepareSplitLogical(DMContext & dm_context,
+                                                               const ColumnDefinesPtr & /*schema_snap*/,
+                                                               const SegmentSnapshotPtr & segment_snap,
+                                                               Handle                     split_point,
+                                                               WriteBatches &             wbs) const
 {
     LOG_INFO(log, "Segment [" << segment_id << "] prepare split logical start");
 
@@ -782,13 +786,13 @@ Segment::SplitInfo Segment::prepareSplitLogical(DMContext & dm_context,
 
     LOG_INFO(log, "Segment [" << segment_id << "] prepare split logical done");
 
-    return SplitInfo(true, split_point, my_stable, other_stable);
+    return {SplitInfo{true, split_point, my_stable, other_stable}};
 }
 
-Segment::SplitInfo Segment::prepareSplitPhysical(DMContext &                dm_context,
-                                                 const ColumnDefinesPtr &   schema_snap,
-                                                 const SegmentSnapshotPtr & segment_snap,
-                                                 WriteBatches &             wbs) const
+std::optional<Segment::SplitInfo> Segment::prepareSplitPhysical(DMContext &                dm_context,
+                                                                const ColumnDefinesPtr &   schema_snap,
+                                                                const SegmentSnapshotPtr & segment_snap,
+                                                                WriteBatches &             wbs) const
 {
     LOG_INFO(log, "Segment [" << segment_id << "] prepare split physical start");
 
@@ -797,7 +801,7 @@ Segment::SplitInfo Segment::prepareSplitPhysical(DMContext &                dm_c
     auto read_info       = getReadInfo(dm_context, *schema_snap, segment_snap);
     auto split_point_opt = getSplitPointSlow(dm_context, read_info, segment_snap);
     if (!split_point_opt.has_value())
-        return SplitInfo();
+        return {};
 
     auto split_point = split_point_opt.value();
 
@@ -869,7 +873,7 @@ Segment::SplitInfo Segment::prepareSplitPhysical(DMContext &                dm_c
 
     LOG_INFO(log, "Segment [" << segment_id << "] prepare split physical done");
 
-    return SplitInfo(false, split_point, my_new_stable, other_stable);
+    return {SplitInfo{false, split_point, my_new_stable, other_stable}};
 }
 
 SegmentPair Segment::applySplit(DMContext &                dm_context, //

--- a/dbms/src/Storages/DeltaMerge/Segment.h
+++ b/dbms/src/Storages/DeltaMerge/Segment.h
@@ -66,20 +66,11 @@ public:
 
     struct SplitInfo
     {
-        bool ok;
-
         bool   is_logical;
         Handle split_point;
 
         StableValueSpacePtr my_stable;
         StableValueSpacePtr other_stable;
-
-        SplitInfo() : ok(false) {}
-
-        SplitInfo(bool is_logical_, Handle split_point_, StableValueSpacePtr my_stable_, StableValueSpacePtr other_stable_)
-            : ok(true), is_logical(is_logical_), split_point(split_point_), my_stable(my_stable_), other_stable(other_stable_)
-        {
-        }
     };
 
     Segment(const Segment &) = delete;
@@ -143,11 +134,11 @@ public:
     /// For those split, merge and mergeDelta methods, we should use prepareXXX/applyXXX combo in real production.
     /// split(), merge() and mergeDelta() are only used in test cases.
 
-    SegmentPair split(DMContext & dm_context, const ColumnDefinesPtr & schema_snap) const;
-    SplitInfo   prepareSplit(DMContext &                dm_context,
-                             const ColumnDefinesPtr &   schema_snap,
-                             const SegmentSnapshotPtr & segment_snap,
-                             WriteBatches &             wbs) const;
+    SegmentPair              split(DMContext & dm_context, const ColumnDefinesPtr & schema_snap) const;
+    std::optional<SplitInfo> prepareSplit(DMContext &                dm_context,
+                                          const ColumnDefinesPtr &   schema_snap,
+                                          const SegmentSnapshotPtr & segment_snap,
+                                          WriteBatches &             wbs) const;
     SegmentPair
     applySplit(DMContext & dm_context, const SegmentSnapshotPtr & segment_snap, WriteBatches & wbs, SplitInfo & split_info) const;
 
@@ -249,15 +240,15 @@ private:
     /// Only look up in the stable vs.
     std::optional<Handle> getSplitPointFast(DMContext & dm_context, const StableSnapshotPtr & stable_snap) const;
 
-    SplitInfo prepareSplitLogical(DMContext &                dm_context, //
-                                  const ColumnDefinesPtr &   schema_snap,
-                                  const SegmentSnapshotPtr & segment_snap,
-                                  Handle                     split_point,
-                                  WriteBatches &             wbs) const;
-    SplitInfo prepareSplitPhysical(DMContext &                dm_context,
-                                   const ColumnDefinesPtr &   schema_snap,
-                                   const SegmentSnapshotPtr & segment_snap,
-                                   WriteBatches &             wbs) const;
+    std::optional<SplitInfo> prepareSplitLogical(DMContext &                dm_context, //
+                                                 const ColumnDefinesPtr &   schema_snap,
+                                                 const SegmentSnapshotPtr & segment_snap,
+                                                 Handle                     split_point,
+                                                 WriteBatches &             wbs) const;
+    std::optional<SplitInfo> prepareSplitPhysical(DMContext &                dm_context,
+                                                  const ColumnDefinesPtr &   schema_snap,
+                                                  const SegmentSnapshotPtr & segment_snap,
+                                                  WriteBatches &             wbs) const;
 
 
     /// Make sure that all delta packs have been placed.

--- a/dbms/src/Storages/DeltaMerge/tests/gtest_dm_segment.cpp
+++ b/dbms/src/Storages/DeltaMerge/tests/gtest_dm_segment.cpp
@@ -724,6 +724,26 @@ try
 }
 CATCH
 
+TEST_F(Segment_test, SplitFail)
+try
+{
+    const size_t num_rows_write = 100;
+    {
+        // write to segment
+        Block block = DMTestEnv::prepareSimpleWriteBlock(0, num_rows_write, false);
+        segment->write(dmContext(), std::move(block));
+    }
+
+    // Remove all data
+    segment->write(dmContext(), HandleRange(0, 100));
+    segment->flushCache(dmContext());
+
+    auto [a, b] = segment->split(dmContext(), tableColumns());
+    EXPECT_EQ(a, SegmentPtr{});
+    EXPECT_EQ(b, SegmentPtr{});
+}
+CATCH
+
 TEST_F(Segment_test, Restore)
 try
 {

